[Bugfix] Resource:take only works at beginning of the game

jhoffmann2 opened the pull request:
Resource:take expects there to always be a pile of resources in the supply and for that pile to always have the guid. This isn't the case because restacking the resources in a different order will change their guid. Instead of guid, we should use the location of the supply tiles to find the resources. This code uses a raycast since it can find the topmost resource in the pile even if they're not stacked

Quinnsicle commented on src/Resource.lua:
```diff
@@ -39,11 +38,39 @@ local Resource = {
+        end
+    end
+
+    if result.getQuantity() ~= -1 then
```
Will this result in an error when trying to find the getQuantity function if result is nil? Consider adding an early return if the result is nil.

Quinnsicle commented:
Looks good other than the potential error I mentioned. I really like the raycast idea, I've been using scripting zones for similar things, but I think a raycast is more maintainable.  This has been on my todo list for a while, but I ran into problems with trying to support cartel cards. This is a good start at improving the resource management, we can handle the cartels later.

jhoffmann2 commented:
ok I addressed your comment. Raycast is great but I think it's a bit slower for situations where you need to query the objects in an area really often. In this case I think it makes sense though because the supply tiles may get moved by players (for example to put them on the cartel cards) so scripting zones wouldn't work great since we'd have to update the position of the zone every time the tiles move
